### PR TITLE
🌱 Update golangci-lint and Go version for golangci-lint workflow

### LINKED ISSUE #6350: gci linter no longer autofixing import orders
gci linter is no longer fixing import order with `make lint-fix`. This issue was noted in #6348. The errors from the linter are opaque making fixing import order a manual and time consuming process for maintainers - and very discouraging for all contributors.

It seems from here that the autofix was disabled in a recent version c/f https://github.com/golangci/golangci-lint/issues/2604. I'm still researching why this occurred and will update with any information here.

We can either 
- remove `gci` from our golangci-lint. 
- or  downgrade to an older version, i.e. roll back #6335 

Until the linter's behaviour in autofixing comes back in line with expectations. I'm in favour of doing a downgrade - gci was really useful when `make lint-fix` was working.

/kind bug
[One or more /area label. See https://github.com/kubernetes-sigs/cluster-api/labels?q=area for the list of labels]

### PR DESCRIPTION
**Pull request by oscr**
**What this PR does / why we need it**:

Updates the GitHub workflow to use Go 1.18 and bumps golangci-lint 1.44.0 -> 1.48.0. Also fixes findings by the upgraded linter.

Reference sources:
https://github.com/golangci/golangci-lint/releases
https://github.com/golangci/golangci-lint/pull/2438

Part of #5968
Fixes #6350

### CODE REVIEW

**Comment by chrischdi** [comment]
We need a version which includes https://github.com/kubernetes-sigs/cluster-api/issues/6350#issuecomment-1141383233

/hold

File: .github/workflows/golangci-lint.yml [hunk 1]
```diff
@@ -2,6 +2,8 @@ name: golangci-lint
 on:
   pull_request:
     types: [opened, edited, synchronize, reopened]
```
Comment (sbueringer) [inline .github/workflows/golangci-lint.yml:4]: There are a few more 1.17's in .golangci.yml

(but we have to wait with merge until the next golangci-lint release)

Comment (oscr) [reply .github/workflows/golangci-lint.yml:4]: Thank you! It's been updated.

**Comment by oscr** [comment]
> @oscr Please add the following to the PR description
> 
> Part of #5968 Fixes #6350
> 
> (2nd one is true after the next golangci-lint release)

Sure! I will bump the golangci-lint version as soon as it's available.

**Comment by sbueringer** [comment]
Perfect. Thank you!

lgtm so far

/ok-to-test

**Comment by killianmuldoon** [comment]
@oscr It would be good to dedupe this with #6943 - would you mind picking up tracking the linter update here as we move toward finally being able to update? 😆

**Comment by killianmuldoon** [comment]
Summary of learnings from #6943:

v1.47.1 does not work due to performance issues in the revive linter. v1.47.2 disables the problematic linters allowing that version of golangci-lint to be used on CAPI at the cost of some linter coverage. 

At this point we could:
- Delay update for now and see if the bug can be fixed in the next couple of weeks
- Update but keep the exclusions in the config so we can roll them back once a fix is out
@fabriziopandini WDYT?

**Comment by chrischdi** [comment]
Relevant for discussing if to delay or not: https://github.com/golangci/golangci-lint/issues/2997#issuecomment-1190681977

List of disabled linters:

> * `var-declaration` (_default_): https://revive.run/r#var-declaration
> * `errorf` (_default_): https://revive.run/r#errorf
> * `unexported-return` (_default_): https://revive.run/r#unexported-return
> * `time-naming` (_default_): https://revive.run/r#time-naming
> * `context-keys-type` (_default_): https://revive.run/r#context-keys-type
> * `modifies-value-receiver`: https://revive.run/r#modifies-value-receiver
> * `range-val-address`: https://revive.run/r#range-val-address
> * `string-of-int`: https://revive.run/r#string-of-int
> * `time-equal`: no documentation
> * `unhandled-error`: https://revive.run/r#unhandled-error

**Comment by oscr** [comment]
I just want to highlight a change. When I cherry picked the changes from #6943 I got some `nolintlint` findings: 

```
util/defaulting/defaulting.go:28:38: directive `//nolint:revive` is unused for linter "revive" (nolintlint)
type DefaultingValidator interface { //nolint:revive
                                     ^
cmd/clusterctl/client/alpha/rollout_pauser.go:39:92: directive `//nolint:revive // MachineDeployment is intentionally capitalized.` is unused for linter "revive" (nolintlint)
			return errors.Errorf("MachineDeploymet is already paused: %v/%v\n", ref.Kind, ref.Name) //nolint:revive // MachineDeployment is intentionally capitalized.
			                                                                                        ^
cmd/clusterctl/client/alpha/rollout_resumer.go:39:99: directive `//nolint:revive // MachineDeployment is intentionally capitalized.` is unused for linter "revive" (nolintlint)
			return errors.Errorf("MachineDeployment is not currently paused: %v/%v\n", ref.Kind, ref.Name) //nolint:revive // MachineDeployment is intentionally capitalized.
			                                                                                               ^
controlplane/kubeadm/internal/etcd/fake/client.go:26:30: directive `//nolint:revive` is unused for linter "revive" (nolintlint)
type FakeEtcdClient struct { //nolint:revive
                             ^
cmd/clusterctl/internal/test/fake_proxy.go:50:35: directive `//nolint:revive` is unused for linter "revive" (nolintlint)
	FakeScheme = runtime.NewScheme() //nolint:revive
```

So I temporary allowed unused in `nolintlint` until revive is fully enabled again. 

```
nolintlint:
    allow-unused: true # TODO(oscr) Set to false when revive linter is fully enabled again.` 
```

I this ok?

**Comment by killianmuldoon** [comment]
> So I temporary allowed unused in nolintlint until revive is fully enabled again.
I this ok?

Yeah - that's perfectly fine I think.

**Comment by killianmuldoon** [comment]
Part of #5968

**Comment by oscr** [comment]
There are now two new versions of `golangci-lint`:
https://github.com/golangci/golangci-lint/releases/tag/v1.47.2
https://github.com/golangci/golangci-lint/releases/tag/v1.47.3

`1.47.2` which just disables the slow rules and `1.47.3` with some upgrades and fixes. 

What would we like to do here? I can upgrade, stay or wait whichever works best.

**Comment by stmcginnis** [comment]
> What would we like to do here? I can upgrade, stay or wait whichever works best.

Probably good to update to 1.47.3 at this point. Looks like it picks up some good fixes.

**Comment by killianmuldoon** [comment]
golangci-lint 1.48.0 is now released, and [it looks like there's not going to be another release for a little while](https://github.com/golangci/golangci-lint/issues/2997#issuecomment-1216830262).

I think we should update CAPI to golangci-lint 1.48.0. I ran it locally and there's no big issues IMO - there's a bunch of new fixes, but they're minor formatting things for the most part. We can also bump Go to 1.18 in the golangci-lint job and in the .golangci.yml

We should track the disabled linters to make sure we're not causing too much drift for when we bring the full range of revive linters back with the release of 1.49.

**Comment by sbueringer** [comment]
Sounds good to me. Additional 1.48 also supports Go 1.19 which unblocks upcoming work for Kubernetes v1.25.0.

**Comment by oscr** [comment]
I upgraded the golangci-lint 1.48 but there is a problem with the license headers. Unfortunately the new version reports the header as a finding. For example:

```
cmd/clusterctl/api/v1alpha3/metadata_type_test.go:8: File is not `gofmt`-ed with `-s` (gofmt)
    http://www.apache.org/licenses/LICENSE-2.0
```
Running `make lint-fix` changes them and then the `verify-boilerplate.sh` script fails. Looking at it.

**Comment by killianmuldoon** [comment]
>Running make lint-fix changes them and then the verify-boilerplate.sh script fails. Looking at it.

I've also been trying to pin this down on https://github.com/kubernetes-sigs/cluster-api/pull/7066/files#diff-c5a9f9e91fea8f113dbe8cc4feece40bed422499e3372094ae037f119b94ff78 

I'll share anything I find - it's down to tabs vs spaces on some files. I'm not sure yet if there's a clean way to fix it. You can run the verify-boilerplate.py script directly (with -v) which might help see what the errors are

**Comment by killianmuldoon** [comment]
I managed to resolve the header issues over at:
https://github.com/killianmuldoon/cluster-api/pull/31

I didn't need to change the boilerplate there's just a small number of the files that have a special character in the header (which some mix of gofmt and verify-boilerplate.sh is catching now).

I just copied the header from one of the files that's passing verify-boilerplate and it's all passing (at least locally 😆 )

**Comment by oscr** [comment]
It seems like it needs a space after the header and before the package. That makes all the difference. 

```
diff --git a/main.go b/main.go
index ba81125db..2445aef8d 100644
--- a/main.go
+++ b/main.go
@@ -13,6 +13,7 @@ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 See the License for the specific language governing permissions and
 limitations under the License.
 */
+
 package main
 
 import (
```

I've added it and now it passed both locally at least

File: main.go [hunk 2]
```diff
@@ -13,6 +13,7 @@ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 See the License for the specific language governing permissions and
 limitations under the License.
 */
+
```
Comment (killianmuldoon) [inline main.go:16]: 🤯

**Review by killianmuldoon — COMMENTED** [review]
Could you also bump the go versions in .golangci.yml? I think it's in there twice.

**Review by killianmuldoon — APPROVED** [review]
/lgtm

File: .github/workflows/golangci-lint.yml [hunk 3]
```diff
@@ -21,9 +21,9 @@ jobs:
       - uses: actions/checkout@v3
       - uses: actions/setup-go@v3
         with:
-          go-version: 1.17
+          go-version: 1.19
```
Comment (sbueringer) [inline .github/workflows/golangci-lint.yml:24]: I would prefer staying on Go 1.18 for the moment until we upgrade everything else to the next Kubernetes and Go version. (applies to all occurences of Go 1.19 on this PR)

Otherwise the linter in CI would be the only thing using Go 1.19 while every other job and we locally would use Go 1.18

@killianmuldoon WDYT?

File: .github/workflows/golangci-lint.yml [hunk 4]
```diff
@@ -21,9 +21,9 @@ jobs:
       - uses: actions/checkout@v3
```
Comment (sbueringer) [inline .github/workflows/golangci-lint.yml:21]: (just had to put the comment somewhere)

Can we please also revert the changes from this [PR](https://github.com/kubernetes-sigs/cluster-api/pull/6351/files) to `internal/webhooks/cluster_test.go` and `internal/webhooks/clusterclass_test.go`

File: .github/workflows/golangci-lint.yml (hunk 3 above)
Comment (oscr) [reply .github/workflows/golangci-lint.yml:24]: From the release notes for 1.19:

```
Changes to the language

There is only one small change to the language, a [very small correction](https://github.com/golang/go/issues/52038) to the [scope of type parameters in method declarations](https://tip.golang.org/ref/spec#Declarations_and_scope). Existing programs are unaffected.
```
source: https://tip.golang.org/doc/go1.19

Comment (oscr) [reply .github/workflows/golangci-lint.yml:24]: My interpretation is that this only affects generic functions and since generics isn't used yet it should be fine.

Comment (sbueringer) [reply .github/workflows/golangci-lint.yml:24]: But it still feels wrong to me if we run golangci-lint in GitHub actions with Go 1.19 while the everything else uses Go 1.18.

E.g. when I run the linter on my machine I will use Go 1.18 as we use 1.18 for everything

File: .github/workflows/golangci-lint.yml (hunk 4 above)
Comment (oscr) [reply .github/workflows/golangci-lint.yml:21]: @sbueringer I reverted and changed the code since the function was moved in this commit: https://github.com/kubernetes-sigs/cluster-api/commit/ce89e60370ba49bd3599f72d80920ac95bc0f994 I hope this is what you wanted.

File: .github/workflows/golangci-lint.yml (hunk 3 above)
Comment (oscr) [reply .github/workflows/golangci-lint.yml]: No problem then! Fixed.

**Comment by oscr** [comment]
Sorry for the spam.

File: .github/workflows/golangci-lint.yml (hunk 4 above)
Comment (sbueringer) [reply .github/workflows/golangci-lint.yml:21]: Looks perfect. Thx!

File: .github/workflows/golangci-lint.yml (hunk 3 above)
Comment (sbueringer) [reply .github/workflows/golangci-lint.yml]: Thx!

**Comment by sbueringer** [comment]
/lgtm

cc @killianmuldoon for a final look

**Review by killianmuldoon — APPROVED** [review]
/lgtm

Thanks again for following this one through @oscr!

**Comment by sbueringer** [comment]
Yup thank you very much!

/approve
